repository/src/interpret/works.rs

```diff
@@ -0,0 +1,337 @@
+            Ok(x) => x,
+            Err(CommitError::Commit(error, commit)) => {
+                return Ok(Err(format!("failed to parse commit {commit}: {error}")));
+            }
+            Err(e) => return Err(e.into()),
+        };
+        for (commit, commit_hash) in &commits {
+            if let Err(e) = csv.apply_commit(commit) {
+                return Ok(Err(format!(
+                    "commit sequence verification failed: {e} at {commit_hash}",
+                )));
+            }
+        }
+
+        let (last_commit, last_commit_hash) = commits.last().expect(
+            "already checked that the received commit is not same as the last finalized block",
+        );
+        if let Commit::Block(_) = last_commit {
+            if csv
+                .verify_last_header_finalization(&last_finalization_proof.proof)
+                .is_err()
+            {
+                return Ok(Err(
+                    "finalization proof is invalid for the last block.".to_owned()
+                ));
+            }
+            advance_finalized_branch(raw, *last_commit_hash, last_finalization_proof).await?;
+        } else {
+            return Ok(Err("fp commit must be on top of a block commit.".to_owned()));
+        }
+    }
+    // If the branch ends with a block, agenda, or agenda proof commit
+    else {
+        if tip_commit_hash == lfi.commit_hash {
+            return Ok(Err("the received commit is already finalized.".to_owned()));
+        }
+
+        // Read the commits in the branch and verify them
+        let commits = match read_commits(raw, lfi.commit_hash, tip_commit_hash).await {
+            Ok(x) => x,
+            Err(CommitError::Commit(error, commit)) => {
+                return Ok(Err(format!("failed to parse commit {commit}: {error}",)));
+            }
+            Err(e) => return Err(e.into()),
+        };
+        for (commit, commit_hash) in &commits {
+            if let Err(e) = csv.apply_commit(commit) {
+                return Ok(Err(format!(
+                    "commit sequence verification failed: {e} at {commit_hash}",
+                )));
+            }
+        }
+
+        // If the commit sequence contains block commit(s) that can be finalized
+        let headers = csv.get_block_headers();
+        if headers.len() > 2 {
+            let (last_header, _) = headers.last().expect(
+                "already checked that the received commit is not same as the last finalized block",
+            );
+            let (second_to_last_header, index) = headers[headers.len() - 2].clone();
+            advance_finalized_branch(
+                raw,
+                commits[index].1,
+                LastFinalizationProof {
+                    height: second_to_last_header.height,
+                    proof: last_header.prev_block_finalization_proof.clone(),
+                },
+            )
+            .await?;
+        }
+
+        // Create the associate branch
```

Comment:
```suggestion
        // Create a branch associated to the last commit
```
It seems as if there is a term "associate branch" so how about fixing it like this?